lxd/storage/drivers/driver_btrfs.go

```diff
@@ -0,0 +1,740 @@
+}
+
+// DeleteVolume deletes a volume of the storage device. If any snapshots of the volume remain then
+// this function will return an error.
+func (d *btrfs) DeleteVolume(volType VolumeType, volName string, op *operations.Operation) error {
+	snapshots, err := d.VolumeSnapshots(volType, volName, op)
+	if err != nil {
+		return err
+	}
+
+	if len(snapshots) > 0 {
+		return fmt.Errorf("Cannot remove a volume that has snapshots")
+	}
+
+	volPath := GetVolumeMountPath(d.name, volType, volName)
+
+	if !shared.PathExists(volPath) || !isBtrfsSubvolume(volPath) {
+		return nil
+	}
+
+	err = btrfsSubvolumeDelete(volPath)
+	if err != nil {
+		return err
+	}
+
+	// Although the volume snapshot directory should already be removed, lets remove it here
+	// to just in case the top-level directory is left.
+	err = deleteParentSnapshotDirIfEmpty(d.name, volType, volName)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// MountVolume simulates mounting a volume. As dir driver doesn't have volumes to mount it returns
+// false indicating that there is no need to issue an unmount.
+func (d *btrfs) MountVolume(volType VolumeType, volName string, op *operations.Operation) (bool, error) {
+	return false, nil
+}
+
+// MountVolumeSnapshot sets up a read-only mount on top of the snapshot to avoid accidental modifications.
+func (d *btrfs) MountVolumeSnapshot(volType VolumeType, volName, snapshotName string, op *operations.Operation) (bool, error) {
+	return false, ErrNotImplemented
+}
+
+// UnmountVolume simulates unmounting a volume. As dir driver doesn't have volumes to unmount it
+// returns false indicating the volume was already unmounted.
+func (d *btrfs) UnmountVolume(volType VolumeType, volName string, op *operations.Operation) (bool, error) {
+	return false, nil
+}
+
+// UnmountVolumeSnapshot removes the read-only mount placed on top of a snapshot.
+func (d *btrfs) UnmountVolumeSnapshot(volType VolumeType, volName, snapshotName string, op *operations.Operation) (bool, error) {
+	return false, nil
+}
+
+// SetVolumeQuota sets the quota on the volume.
+func (d *btrfs) SetVolumeQuota(volType VolumeType, volName, size string, op *operations.Operation) error {
+	volPath := GetVolumeMountPath(d.name, volType, volName)
+
+	volID, err := d.getVolID(volType, volName)
+	if err != nil {
+		return err
+	}
+
+	return d.setQuota(volPath, volID, size)
+}
+
+// quotaProjectID generates a project quota ID from a volume ID.
+func (d *btrfs) quotaProjectID(volID int64) uint32 {
+	return uint32(volID + 10000)
+}
+
+// initQuota initialises the project quota on the path. The volID generates a quota project ID.
+func (d *btrfs) initQuota(path string, volID int64) error {
+	if volID == 0 {
+		return fmt.Errorf("Missing volume ID")
+	}
+
+	ok, err := quota.Supported(path)
+	if err != nil || !ok {
+		// Skipping quota as underlying filesystem doesn't suppport project quotas.
+		return nil
+	}
+
+	err = quota.SetProject(path, d.quotaProjectID(volID))
```

Comment:
@monstermunchkin is project quota used with btrfs? I thought this was only used with ext4.

I may well be wrong though, @stgraber is that right?

Reply:
You're right.